### PR TITLE
Add an interface to FTSW500

### LINKED ISSUE #322: Add support for interfacing with ABB's spectrometer software
ABB's software is available in the RDS along with some documentation and a program for generating simulated spectrometer signals. It can be controlled via a simple TCP-based protocol. In addition to being able to start and stop recording via this protocol, it seems that a fairly large proportion of the program's functionality is also exposed this way (more so than the Bruker program used before).

### PR DESCRIPTION
**Pull request by dc2917**
# Description

This PR adds an interface to ABB's FTSW500 software.

A new spectrometer interface base device, `FTSW500InterfaceBase`, is added, which inherits from the abstract `SpectrometerBase` class. This provides the core `connect`, `start_measuring` and `stop_measuring` commands, which are executed when the corresponding buttons on the GUI are pressed.

The `FTSW500Interface` provides the underlying functionality for sending instructions to FTSW500, as well as receiving and handling any responses. The status of FTSW500 is translated into a `SpectrometerStatus` which largely provides a compatible set of states. The main difference is that once FTSW500 connects to an instrument, it begins acquiring data, but requires further instruction to write acquired data to file. This is in contrast to the interface to the EM27, which always writes acquired data to file, but does not begin acquiring until told to. (We could, of course, mimic this behaviour with FTSW500 by effectively making `connect` redundant, then having `start_measuring` perform `connect` + `start measuring`.)

I'm sure some of this can be tidied up but I thought I'd get some feedback first.

Fixes #322 

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [ ] Optimisation (non-breaking, back-end change that speeds up the code)

## Key checklist

- [X] Pre-commit hooks run successfully (`pre-commit run -a`)
- [X] All tests pass (`pytest`)
- [X] The documentation builds without warnings (`mkdocs build -s`)
- [X] Check the GUI still works (if relevant)
- [ ] Check the code works with actual hardware (if relevant)
- [ ] Check the `pyinstaller`-built executable works (if relevant)

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests have been added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

File: finesse/config.py [hunk 1]
```diff
@@ -76,6 +76,18 @@
...
 """
 
+FTSW500_HOST = "127.0.0.1"
+"""The IP address or hostname of the machine running the FTSW500 software."""
+
+FTSW500_PORT = 7778
+"""The port on which the TCP server of FTSW500 is listening."""
+
+FTSW500_POLLING_INTERVAL = 1.0
+"""How long to wait between polls of the EM27's status."""
```
Comment (jamesturner246) [inline finesse/config.py:86]: Is this related to `EM27_SENSORS_POLL_INTERVAL` above? Could you just use that instead? Or maybe is `EM27` a typo?

File: finesse/hardware/plugins/spectrometer/dummy_ftsw500_interface.py
Comment (jamesturner246) [inline finesse/hardware/plugins/spectrometer/dummy_ftsw500_interface.py:1]: Is the dummy interface coming in another PR?

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                return None
+        else:
+            return None
+    elif "ACK" in msg:
+        if "&" in msg:
+            try:
+                status = int(msg.split("&")[1])
+            except ValueError:
+                logging.info(f"{msg.split('&')[1][:-1]}")
+                return None
```
Comment (jamesturner246) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:67]: Is it possible that multiple string arguments are returned? This code wouldn't pick up any after the first.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        (less than 500 ms) or when the FTSW500_SDK object is not well initialized
+
+    Args:
+        response: the byte sequence received from FTSW500
+
+    Returns:
+        SpectrometerStatus: the generic spectrometer device state of FTSW500
+    """
+    msg = response.decode()
+    if "NAK" in msg:
```
Comment (jamesturner246) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:52]: I think (personally) the following makes it clearer what a message looks like. Feel free to ignore -- kinda subjective.
```suggestion
    if msg[:3] == "NAK"
```

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        if "&" in msg:
+            try:
+                status = int(msg.split("&")[1])
+            except ValueError:
+                logging.error(f"{msg.split('&')[1][:-1]}")
+                return None
+        else:
+            return None
+    elif "ACK" in msg:
```
Comment (jamesturner246) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:61]: Same here:
```suggestion
    elif msg[:3] == "ACK":
```

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            except ValueError:
+                logging.info(f"{msg.split('&')[1][:-1]}")
+                return None
+        else:
+            return None
+    else:
+        raise FTSW500Error("Unrecognised response")
+
+    if status == -1:
+        return SpectrometerStatus(1)
```
Comment (jamesturner246) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:74]: How come -1 changed to 1? I guess an intermediate state is kind of like connecting.

**Review by jamesturner246 — CHANGES_REQUESTED** [review]
Looking good. 

I put some questions below, and a kinda subjective suggestion that I'm not worried about being ignored.

Would be good to have some unit tests for this class though. Presumably they'd be very similar to the other spectrometer class.

File: finesse/hardware/plugins/spectrometer/dummy_ftsw500_interface.py
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/dummy_ftsw500_interface.py:1]: I'm not sure what to do about this. I'm not sure it's needed, as the dummy OPUS is generic enough that it's essentially just a dummy spectrometer.

File: finesse/config.py (hunk 1 above)
Comment (dc2917) [reply finesse/config.py:86]: Good spot 😅

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 3 above)
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:52]: Yeah, good idea

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        if "&" in msg:
+            try:
+                status = int(msg.split("&")[1])
+            except ValueError:
+                logging.error(f"{msg.split('&')[1][:-1]}")
+                return None
+        else:
+            return None
+    elif "ACK" in msg:
+        if "&" in msg:
```
[the remark this answers is not in the document]
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:62]: It shouldn't appear anywhere else, no. Perhaps it's safer to look for `ACK&` and `NAK&`?

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        response: the byte sequence received from FTSW500
+
+    Returns:
+        SpectrometerStatus: the generic spectrometer device state of FTSW500
+    """
+    msg = response.decode()
+    if "NAK" in msg:
+        if "&" in msg:
+            try:
+                status = int(msg.split("&")[1])
```
Comment (jamesturner246) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:55]: If it's parsing `msg.split("&")[1] == 'false'` or `true`, wouldn't the integer cast fail?

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 5 above)
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:74]: This is because `SpectrometerStatus` is based on EM27's statuses, so this just converts a FTSW500 status to a(n apparently) generic spectrometer status

File: finesse/hardware/plugins/spectrometer/dummy_ftsw500_interface.py
Comment (jamesturner246) [reply finesse/hardware/plugins/spectrometer/dummy_ftsw500_interface.py:1]: Ah, fair enough. Can probably just remove this then, right?

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:62]: Ah yes, just in case `msg` contains `ACK` or `NAK`, right? true.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 5 above)
Comment (jamesturner246) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:74]: I see. okay.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 7 above)
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:55]: Yeah the `except ValueError:` is there to catch non-integer messages. This should only be used for obtaining the status, so if there's a non-integer message something must have gone awry, so it's sent to the logger

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 3 above)
Comment (dalonsoa) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:52]: What about:
```suggestion
    if msg.startswith("NAK"):
```

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 4 above)
Comment (dalonsoa) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:61]: Indeed...
```suggestion
    elif msg.startswith("ACK"):
```

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 7 above)
Comment (jamesturner246) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:55]: Aahhh, I see... Maybe add a comment or something to that effect.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:62]: `msg` will always contain "ACK" or "NAK", and the "&" we want to search for is a delimiter which should always come after the "ACK" or "NAK". So by instead searching for "ACK&" and "NAK&" we can be sure that we're getting the delimiting "&".

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 5 above)
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:74]: Yeah, it felt a bit weird to do. I felt the base spectrometer device framework a bit restrictive but thought I'd see what I could do without changing it.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 3 above)
Comment (jamesturner246) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:52]: There really is a Python method for everything!

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    Returns:
+        SpectrometerStatus: the generic spectrometer device state of FTSW500
+    """
+    msg = response.decode()
+    if msg.startswith("NAK"):
+        if "&" in msg:
+            try:
+                status = int(msg.split("&")[1])
+            except ValueError:
+                logging.error(f"{msg.split('&')[1][:-1]}")
```
Comment (dalonsoa) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:57]: Maybe add something more explicit like: 
```suggestion
                logging.error(f"Error while reading from FTSW500: {msg.split('&')[1][:-1]}")
```
so the user knows where the error is coming from.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dalonsoa) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:62]: If the code should only do something special (besides returning `None`) if there is `ACK&` or `NAK&`, then I would suggest to just look for that string (in particular using the `startswith` method)

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+            self._requester.recv(1024)
+
+    def _regurgitate_nonmodal_dialog_message(self) -> None:
+        """Obtain the last non-modal message displayed on FTSW500 and log it."""
+        if self._check_is_nonmodal_dialog_open():
+            self._requester.sendall(b"getLastNonModalMessageDisplayed\n")
+            data = self._requester.recv(1024)
+            logging.info(f"FTSW500: {data.decode().split('&')[1][:-1]}")
+            self._requester.sendall(b"closeNonModalDialogMessage\n")
+            self._requester.recv(1024)
```
Comment (dalonsoa) [inline finesse/hardware/plugins/spectrometer/ftsw500_interface.py:142]: While the messages sent are different, the logic in both cases seems to be the same. I'd suggest you combine both and use wrappers for each of them. Something like:

```suggestion
    def _is_dialog_open(self, check_modal: bool) -> bool:
        """Query whether FTSW500 has a dialog open."""
        msg = "" if check_modal else "Non"
        self._requester.sendall(f"is{msg}ModalMessageDisplayed\n".encode("utf-8"))
        data = self._requester.recv(1024)
        if data != b"" and data.decode().split("&")[1] == "true\n":
            return True
        return False
        
    def _regurgitate_dialog_message(self, is_modal: bool) -> None:
        """Obtain the last message displayed on FTSW500 and log it."""
        msg = "" if is_modal else "Non"
        self._requester.sendall(f"getLast{msg}ModalMessageDisplayed\n".encode("utf-8"))
        data = self._requester.recv(1024)
        logging.info(f"FTSW500: {data.decode().split('&')[1][:-1]}")
        self._requester.sendall(f"close{msg}ModalDialogMessage\n".encode("utf-8"))
        self._requester.recv(1024)

    def _regurgitate_modal_dialog_message(self) -> None:
        """Obtain the last modal message displayed on FTSW500 and log it."""
        if self._is_dialog_open(check_modal=True)
            self._regurgitate_dialog_message(is_modal=True)

    def _regurgitate_non_modal_dialog_message(self) -> None:
        """Obtain the last non modal message displayed on FTSW500 and log it."""
        if self._is_dialog_open(check_modal=False)
            self._regurgitate_dialog_message(is_modal=False)
```

**Review by dalonsoa — APPROVED** [review]
Keeping aside some tests, this looks good!

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 2 above)
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py:67]: No, there should only ever be one string. Could potentially mess things up if the string contains a `'&'`, but if we instead search for "ACK&" and "NAK&" as suggested [above](https://github.com/ImperialCollegeLondon/FINESSE/pull/521#discussion_r1500560751) we should be able to account for that.

File: finesse/hardware/plugins/spectrometer/ftsw500_interface.py (hunk 9 above)
Comment (dc2917) [reply finesse/hardware/plugins/spectrometer/ftsw500_interface.py]: I'm tempted to remove the handling of modal dialog messages from our interface, as these generally mean user input is required in FTSW500 directly.

**Review by jamesturner246 — APPROVED** [review]
Looks good! Are you planning on adding tests in another PR?
